kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+                final FileStatus nextFile = filesIter.next();
+                final Path nextFilePath = new Path(nextFile.getPath());
+
+                try {
+                    if (nextFilePath.getName().endsWith(".json")) {
+                        final JsonHandler jsonHandler = tableClient.getJsonHandler();
+
+                        // Convert the `nextFile` FileStatus into an internal ScanFile Row and then
+                        // allow the connector to contextualize it (i.e. perform any splitting)
+                        final CloseableIterator<FileReadContext> fileReadContextIter =
+                            jsonHandler.contextualizeFileReads(
+                                Utils.singletonCloseableIterator(
+                                    InternalUtils.getScanFileRow(nextFile)),
+                                Literal.TRUE
+                            );
+
+                        // Read that file
+                        final CloseableIterator<FileDataReadResult> fileReadDataIter =
+                            tableClient.getJsonHandler().readJsonFiles(
+                                fileReadContextIter,
+                                readSchema);
+
+                        return combine(fileReadDataIter, false /* isFromCheckpoint */);
+                    } else if (nextFilePath.getName().endsWith(".parquet")) {
+                        final ParquetHandler parquetHandler = tableClient.getParquetHandler();
+
+                        // Convert the `nextFile` FileStatus into an internal ScanFile Row and then
+                        // allow the connector to contextualize it (i.e. perform any splitting)
+                        final CloseableIterator<FileReadContext> fileWithContext =
+                            parquetHandler.contextualizeFileReads(
+                                Utils.singletonCloseableIterator(
+                                    InternalUtils.getScanFileRow(nextFile)),
+                                Literal.TRUE);
+
+                        // Read that file
+                        final CloseableIterator<FileDataReadResult> fileReadDataIter =
+                            tableClient.getParquetHandler().readParquetFiles(
+                                fileWithContext,
+                                readSchema);
+
+                        return combine(fileReadDataIter, true /* isFromCheckpoint */);
+                    } else {
+                        throw new IllegalStateException(
+                            String.format("Unexpected log file path: %s", nextFilePath)
+                        );
+                    }
+                } catch (IOException ex) {
+                    throw new RuntimeException(ex);
+                }
+            }
+        };
+    }
+
+    /**
+     * Take input (iterator<T>, boolean) and produce an iterator<T, boolean>.
+     */
+    private CloseableIterator<Tuple2<FileDataReadResult, Boolean>> combine(
```

Comment:
Is there a way we can get rid of this output `isFromCheckpoint`?

Does this work?
When reading from checkpoint files, just don't read the `remove` column. Just attach a all null vector in place of `remove` column. When iterating check check for non-nulls in `remove` column for creating tombstone file list.

Reply:
```java
// Note: No AddFile will appear twice in a checkpoint, so we only need
//       non-checkpoint AddFiles in the set
if (!isFromCheckpoint) {
    addFilesFromJson.add(key);
}
```

I think we still need it. Perhaps we can revisit this later